HistoryItem are sometimes skipped on back/forward navigation on GitHub.com

cdumez opened the pull request:
#### 165f03cc6003b24c3ddaf46517c3f72183c5f4fe
<pre>
HistoryItem are sometimes skipped on back/forward navigation on GitHub.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=243914">https://bugs.webkit.org/show_bug.cgi?id=243914</a>
&lt;rdar://98392025&gt;

Reviewed by Darin Adler.

In 252649@main, I relaxed the user gesture requirement for our back/forward list
hijacking prevention, by treating a navigation as having a user gesture if there
was user interaction within the last second. This seemed to address issues on
GitHub.com.

However, back/forward navigation on GitHub is still flaky because it sometimes
takes GitHub a little over 1 second to call history.pushState() after the user
has clicked a link. To address this, I am relaxing our heuristic even further
to 10 seconds instead.

Note that Chrome seems to only require having ever interacted with the page
so we&apos;re still stricter.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::hasRecentUserInteractionForNavigationFromJS const):

Canonical link: <a href="https://commits.webkit.org/253405@main">https://commits.webkit.org/253405@main</a>
</pre>

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7268,7 +7268,8 @@ bool Document::hasRecentUserInteractionForNavigationFromJS() const
         return true;
 
-    return (MonotonicTime::now() - lastHandledUserGestureTimestamp()) <= UserGestureToken::maximumIntervalForUserGestureForwarding;
+    static constexpr Seconds maximumItervalForUserGestureForwarding { 10_s };
```
10s seems like a really long time. How about 4s or 5s?

cdumez replied:
I am trying to be conservative since this is new behavior and we are still way stricter than Chrome on this (as mentioned in my ChangeLog). It seems GitHub is waiting on some kind of networking before calling pushState() so I was particularly worried about slow networking. My proposal would be to use 10s for the branch and we can try to lower it on trunk.

cdumez committed:
HistoryItem are sometimes skipped on back/forward navigation on GitHub.com
https://bugs.webkit.org/show_bug.cgi?id=243914
<rdar://98392025>

Reviewed by Darin Adler.

In 252649@main, I relaxed the user gesture requirement for our back/forward list
hijacking prevention, by treating a navigation as having a user gesture if there
was user interaction within the last second. This seemed to address issues on
GitHub.com.

However, back/forward navigation on GitHub is still flaky because it sometimes
takes GitHub a little over 1 second to call history.pushState() after the user
has clicked a link. To address this, I am relaxing our heuristic even further
to 10 seconds instead.

Note that Chrome seems to only require having ever interacted with the page
so we're still stricter.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::hasRecentUserInteractionForNavigationFromJS const):

Canonical link: https://commits.webkit.org/253405@main
